### PR TITLE
drivers: stm32_temp stm32h5 device must disable icache to access cal

### PR DESCRIPTION
**Pull request by tu-brian**
Reading the temperature calibration data requires disabling the icache of the stm32h5x mcu.
Else a bus fault error occurs reading Address: 0x8fff814-0x8fff818.
Enable afterwards.

### CODE REVIEW

**Review by FRASTM — DISMISSED** [review]

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 1]
```diff
@@ -110,6 +118,11 @@ static int stm32_temp_channel_get(const struct device *dev, enum sensor_channel
 	temp /= ((*cfg->cal2_addr - *cfg->cal1_addr) >> cfg->ts_cal_shift);
 #endif
 	temp += cfg->cal1_temp;
+
+#if defined(CONFIG_SOC_SERIES_STM32H5X)
+	LL_ICACHE_Enable();
+#endif /* CONFIG_SOC_SERIES_STM32H5X */
+
```
[the remark this answers is not in the document]
Comment (tu-brian) [reply drivers/sensor/stm32_temp/stm32_temp.c:125]: I Get your point, but for this fix I simply ported what was done in https://github.com/zephyrproject-rtos/zephyr/commit/8d5ff8cb68673eaa0f1023af37148c37ed782bf6 to here. When searching the codebase for LL_CACHE I can't find any other place where the check before enable is implemented

[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/stm32_temp/stm32_temp.c:125]: > I Get your point, but for this fix I simply ported what was done in [8d5ff8c](https://github.com/zephyrproject-rtos/zephyr/commit/8d5ff8cb68673eaa0f1023af37148c37ed782bf6) to here. When searching the codebase for LL_CACHE I can't find any other place where the check before enable is implemented

Ok I see. Sorry for the noise. By the way, another point... This is instruction cache, so what does it mean reading 0x8fff814-0x8fff818 addresses? Is it some code fetched from that location? Can you comment about it a little bit more in the commit header? Sorry again if the question is pointless...

[the remark this answers is not in the document]
Comment (tu-brian) [reply drivers/sensor/stm32_temp/stm32_temp.c:125]: @FRASTM Can you please explain why the fix you did in https://github.com/zephyrproject-rtos/zephyr/commit/8d5ff8cb68673eaa0f1023af37148c37ed782bf6 solved the bus error on reading the read-only area?

**Comment by gautierg-st** [comment]
How did you test that? die_temp is not present (yet) in stm32h5 dtsi. I suppose you have added it to detect this problem but we can't reproduce without it. Could you add it to your PR?

**Comment by tu-brian** [comment]
> How did you test that? die_temp is not present (yet) in stm32h5 dtsi. I suppose you have added it to detect this problem but we can't reproduce without it. Could you add it to your PR?

I tested the implementation on our own board, with the required dts entry added. 
For the sake of this PR, I can add a support to samples/sensor/die_temp_polling. Will that be sufficient?

**Comment by gautierg-st** [comment]
Support for die_temp should be added directly in the stm32h5.dtsi like it is usually done, and the node should be enabled and the alias created on at least one H5 board (I have the stm32h573i_dk to test). If you could add that, I will be able to test on my side.

File: samples/sensor/die_temp_polling/boards/stm32h573i_dk.overlay [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+&die_temp {
+  	status = "okay";
+};
```
Comment (gautierg-st) [inline samples/sensor/die_temp_polling/boards/stm32h573i_dk.overlay:8]: Could you move that directly in the stm32h573i_dk.dts?
Ideally create a separate commit to have one commit for the dtsi and another for the dts, but not required.

**Review by gautierg-st — CHANGES_REQUESTED** [review]
I've tested your PR, and it is indeed necessary to disable the icache to read the calibration values.
Minor change requested, otherwise LGTM.
Thanks

**Review by gautierg-st — APPROVED** [review]

**Review by erwango — APPROVED** [review]

File: drivers/sensor/stm32_temp/stm32_temp.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/sensor/stm32_temp/stm32_temp.c:125]: > @FRASTM Can you please explain why the fix you did in [8d5ff8c](https://github.com/zephyrproject-rtos/zephyr/commit/8d5ff8cb68673eaa0f1023af37148c37ed782bf6) solved the bus error on reading the read-only area?

I got the idea from https://github.com/STMicroelectronics/STM32CubeH5/tree/main/Projects/NUCLEO-H563ZI/Examples_LL/UTILS/UTILS_ReadDeviceInfo

[the remark this answers is not in the document]
Comment (mathieuchopstm) [reply drivers/sensor/stm32_temp/stm32_temp.c:125]: (For future readers)

On STM32H5, the ICACHE block is interposed on C-bus between Cortex-M33 and FLASH. (See RM0492 §2.1.1 `Fast C-bus`)

The ICACHE determines whether an access is cacheable or not based on the AHB attribute, which depends on MPU configuration. (§8.4.6 `Cacheable and non-cacheable traffic`).

When a cacheable transaction is received from the Cortex-M33, if the requested data is not present in cache, a *cache miss* occurs and a cache line refill (128-bit access via burst) is performed (§8.4.7 `Cacheable accesses`).

However, all accesses to OTP and RO regions must be done with caching disabled (§7.3.2 `FLASH signals`); indeed, accesses to this region must be 16 or 32-bit sized, or a bus error is raised (§7.5.9 `One-time-programmable and read-only memory protections` - table 38 `OTP/RO access constraints`).

By disabling the ICACHE, we ensure the manufacturing (RO) flash region is accessed by 16/32-bit size as it should.

**Pull request merged**